### ISSUE TITLE
Configurable JSONP Callback Validator

### ISSUE THREAD

**Issue by jhaber** [FasterXML/jackson-databind#1693]
I wanted to get your thoughts on a new feature/option. Right now `JSONPObject` allows any string for function name and passes it straight through with `writeRaw`. This creates a potential attack vector ([reflected file download](http://blog.davidvassallo.me/2014/11/02/practical-reflected-file-download-and-jsonp/) for example). In most cases you only want to allow valid JavaScript identifiers as function names, not arbitrary JavaScript code. Do you think it would be reasonable to have a mechanism to enforce this within Jackson, or do you think it should always be application's responsibility to perform this validation?

**Comment by cowtowncoder** [comment]
Good question. I suspect that any validation to perform would have to be configurable, or allow disabling; and if added on a minor release would result in a bug report or two. Users tend to be quite creative in finding ways to make use of seemingly useless functionality, such as ability to do more (... damage, potentially :) ) with less restrictions.

One possibility, I think, would be to add this idea as one more thing into hypothetical "Jackson security module". This is something I plan on doing quite soon (but not before 2.9), initially focusing on preventing loading of "known bad classes", above and beyond what databind does by default.
This would both allow faster response times for reported threats (and for wider range of versions), but also let users configure more elaborate setups.

Now... beyond classes (types), this would seem like one obvious thing to add. And if so, could have more complex/powerful configurability; databind's API has to be more constrained because it allows generic overall configuration, but new module can focus more on security aspects.

One obvious challenge, however, would be that whereas such module can work to block loading of classes (via existing extension points), working with JSONP would probably require new ext points. And that can not be done in patches.

Another possibility could be consider addition in JAX-RS module. But even there timing of 2.9 is bit strict at this point.

I guess my main concern really is that of where to put validation, in a way that allows configuration and overrides  -- and more specifically that databind itself is challenging.

As usual I am open to suggestions and ideas. And I think this is very much a valid concern to tackle.

**Comment by jhaber** [comment]
Thanks for the thoughtful response. I'm not sure if it's idiomatic, but one possibility is to register a custom serializer for JSONPObject on the ObjectMapper as a means to implement this validation. There could be a stricter serializer (with various options) that lives in the new security module. If you want this validation, you would configure the strict serializer and register it on your ObjectMapper (or more likely, this would be wrapped in a module). With this approach I don't think you would need to make any changes to databind, but it could be a slightly confusing debugging experience if users don't realize that a custom serializer has been registered (since JSONPObject implements JsonSerializable, people might expect that serialization code to always get called). In the short term, we'll probably do something like this internally.

**Comment by cowtowncoder** [comment]
@jhaber Good point, and yes, you are right -- this is how it could and probably should work.

**Comment by cowtowncoder** [comment]
One quick note: new builder-style construction of `JsonMapper` would make it relatively easy to configure validators.

**Comment by cowtowncoder** [comment]
While this may be a good idea for addition, there are no plans to work on this so will close as part of issue clean up. May be re-opened/re-filed if someone decides to work on something like this.
